test/mitmproxy/proxy/layers/http/test_http2.py

```diff
@@ -780,3 +780,24 @@ def test_request_smuggling_te(tctx):
             << CloseConnection(tctx.client)
     )
     assert b"Connection-specific header field present" in err()
+
+
+def test_request_keepalive(tctx):
+    playbook, cff = start_h2_client(tctx)
+    tctx.options.http2_ping_threshold = 1
+    flow = Placeholder(HTTPFlow)
+    server = Placeholder(Server)
+    initial = Placeholder(bytes)
+    assert (
+            playbook
+            >> DataReceived(tctx.client,
+                            cff.build_headers_frame(example_request_headers, flags=["END_STREAM"]).serialize())
+            << http.HttpRequestHeadersHook(flow)
+            >> reply()
+            << http.HttpRequestHook(flow)
+            >> reply()
+            << OpenConnection(server)
+            >> reply(None, side_effect=make_h2)
+            << RequestWakeup(1)
+            << SendData(server, initial)
```

Comment:
We should emulate a `Wakeup` event here and make sure that it triggers a PING. :-)

```python
def test_request_keepalive(monkeypatch, tctx):
    ...
    assert (playbook ...)
    t = time.time()
    monkeypatch.setattr(time, "time", lambda: t + 60)
    assert (playbook >> Wakeup() ...)
```

Reply:
I'll do this at the very end, when the rest is working. :)